rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,199 @@
+# RFC: Converged component implementation pattern
+
+<!--
+An RFC can be anything. A question, a suggestion, a plan. The purpose of this template is to give some structure to help folks write successful RFCs. However, don't feel constrained by this template; use your best judgement.
+
+Tips for writing a successful RFC:
+
+- Simple plain words that make your point, fancy words obfuscate
+- Try to stay concise, but don't gloss over important details
+- Try to write a neutral problem statement, not one that motivates your desired solution
+- Remember, "Writing is thinking". It's natural to realize new ideas while writing your proposal
+-->
+
+---
+
+_List contributors to the proposal here_
+
+## Summary
+
+<!-- Explain the proposed change -->
+
+Tries to commit the common component pattern to writing in the repo for future reference and track improvements
+
+Code samples included could possibly be applied to utility scripts to generate packages
+
+## Background
+
+<!-- If there is relevant background include it here -->
+
+Many new engineers can find it hard to adapt to a new codebase, especially when the coding patterns and practices are fairly new and in a constant state of flux
+
+## Problem statement
+
+<!--
+Why are we making this change? What problem are we solving? What do we expect to gain from this?
+
+This section is important as the motivation or problem statement is indepenent from the proposed change. Even if this RFC is not accepted this Motivation can be used for alternative solutions.
+
+In the end, please make sure to present a neutral Problem statement, rather than one that motivates a particular solution
+-->
+
+Many new developers might have questions on the patterns that we follow to write a React component. This RFC attempts to commit and possibly improve our code patterns in writing so that there might be a reference for new (and old) members of the team.
+
+## Detailed Design or Proposal
+
+<!-- This is the bulk of the RFC. Explain the proposal or design in enough detail for the inteded audience to understand. -->
+
+### Hook based architecture.
+
+Generally a component will have the following different files. Let us consider a `Sample` component
+
+- Sample.tsx
+- useSample.ts
+- useSampleClasses/useSampleStyles.ts
+- renderSample.tsx
+- Sample.types.ts
+
+#### renderSample.tsx
+
+Renders the correct JSX output of the component and its slots given the correct state.
```

Comment:
It is obvious but I would add a note that this is pure `state -> JSX` function. All React magic happens in `useSample`

Reply:
Do you mean `pure state` as in the output of `useSample` ? 

I have understood that currently, was not aware of any other state floating around

Reply:
I mean `renderSample` is a pure function. Not react hook. You cannot do any React-related things there but rather need to prepare everything into `state` in `useSample`.

Reply:
Yeah gotcha, that makes total sense

Reply:
added: `This should be a pure function whose sole responsibility is to render JSX from the provided state. No state mutation or processing
should happen in this function, but rather done in the `useSample` hook`